src/EFCore.Analyzers/InternalUsageDiagnosticAnalyzer.cs

```
@@ -0,0 +1,131 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.EntityFrameworkCore
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class InternalUsageDiagnosticAnalyzer : DiagnosticAnalyzer
+    {
+        public const string Id = "EF1001";
+
+        public const string MessageFormat
+            = "'{0}' supports the Entity Framework Core infrastructure and is not intended to be used directly from your code. " +
+              "This API may change or be removed in future releases.";
+
+        protected const string DefaultTitle = "Internal EF Core API usage.";
+        protected const string Category = "Usage";
+
+        static readonly int EFLen = "EntityFrameworkCore".Length;
+
+        private static readonly DiagnosticDescriptor _descriptor
+            = new DiagnosticDescriptor(
+                Id,
+                title: DefaultTitle,
+                messageFormat: MessageFormat,
+                category: Category,
+                defaultSeverity: DiagnosticSeverity.Warning,
+                isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(_descriptor);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(AnalyzeNode,
+                SyntaxKind.SimpleMemberAccessExpression,
+                SyntaxKind.ObjectCreationExpression,
+                SyntaxKind.ClassDeclaration);
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
```

Not super-familiar with analyzer code, but this looks good to me.

One suggestion: for the previous analyzer we were forced to patch when an exception escaped from the analyzer code it would crash VS. I'm not sure what best practice is, but can we double check that we don't think this code will throw with anything that VS (and all our creative developers) can throw at it?

---

Good point, syntax trees (and semantic models) tend to be very rich and it's hard to anticipate whether an exception can escape or not. I will confirm what the effect of an exception is on VS, @bricelam I think you worked on the analyzer, any experience/guidance here? We should possibly just swallow exceptions, the effect would only be a missed warning.

---

I was wondering if we should warn if an exception occurs but with a different message that says we hit an unexpected exception and providing a fwlnk to file a bug?

---

I'm not sure what the expected behavior is from an analyzer, I'll do a bit of research on this question tomorrow.

---

> One suggestion: for the previous analyzer we were forced to patch when an exception escaped from the analyzer code it would crash VS.

A bit of research seems to indicate that exceptions thrown from analyzers is *not* supposed to crash VS, but rather cause the analyzer to be disabled: see https://github.com/dotnet/roslyn/issues/3005, https://github.com/dotnet/roslyn/issues/521#issuecomment-74449640 and https://github.com/dotnet/roslyn/issues/268.

I did a quick test and when the analyzer throws an exception, the following appears in the build log:

> 1>CSC : warning AD0001: Analyzer 'Microsoft.EntityFrameworkCore.InternalUsageDiagnosticAnalyzer' threw an exception of type 'System.Exception' with message 'WAT'.

So it seems like the right thing to do is not to swallow - VS catches these exceptions and reports them correctly.

Is there any background on VS crashing with the previous analyzer? @bricelam may you can shed some light on it?

---

Analyzer crashes VS when it runs out of memory. We ran into it in past with infinite recursion https://github.com/aspnet/EntityFrameworkCore/issues/12054

Error in analyzer came up like that only (without crashing VS). We did not make explicit decision on what to do with it so we did just try-catch. https://github.com/aspnet/EntityFrameworkCore/issues/13474

---

Thanks @smitpatel. The new analyzer doesn't do recursion so seems safer in that respect.

Regarding exceptions in general, it seems correct to let them bubble up (rather than swallow as in #13731) - they are cleanly reported in user builds, and users would hopefully file issues for us to fix as in #13474.